cocotb/_sim_versions.py

```diff
@@ -6,39 +6,102 @@
 Classes to compare simulation versions.
 
 These are for cocotb-internal use only.
-
-.. warning::
-    These classes silently allow comparing versions of different simulators.
 """
 
 from distutils.version import LooseVersion
 
 
-class ActivehdlVersion(LooseVersion):
+class _LooseVersionTypeChecked(LooseVersion):
+    """Do a strict type check so that versions of different simulators cannot be compared.
+
+    Example:
+        >>> IcarusVersion("10.3 (stable)") == VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparison not supported between instances of 'IcarusVersion' and 'VerilatorVersion'
+        >>> IcarusVersion("10.3 (stable)") != VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparison not supported between instances of 'IcarusVersion' and 'VerilatorVersion'
+        >>> IcarusVersion("10.3 (stable)") > VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: '>' not supported between instances of 'IcarusVersion' and 'VerilatorVersion'
+        >>> IcarusVersion("10.3 (stable)") >= VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: '>=' not supported between instances of 'IcarusVersion' and 'VerilatorVersion'
+        >>> IcarusVersion("10.3 (stable)") < VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: '<' not supported between instances of 'IcarusVersion' and 'VerilatorVersion'
+        >>> IcarusVersion("10.3 (stable)") <= VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: '<=' not supported between instances of 'IcarusVersion' and 'VerilatorVersion'
+    """
+
+    def __eq__(self, other):
+        if not isinstance(self, type(other)) and not isinstance(other, type(self)):
+            raise TypeError(
+                "Comparison not supported between instances of '{}' and '{}'".format(
+                    type(self).__name__, type(other).__name__
+                )
+            )
+        else:
+            return super().__eq__(other)
```

Comment:
Are we sure we want this overload? This prevents doing semi-reasonable things like having a set of blacklisted versions across simulators and checking membership, as `set.__contains__` will try `__eq__`.

Returning `NotImplemented` as below will make `IcarusVersion("1") == AldecVersion("1")` `False`, which seems quite sensible to me.

Reply:
Unless I'm misunderstanding you, it would also make `IcarusVersion("1") == IcarusVersion("1")` be `False`, wouldn't it?

Reply:
No, it would be True - I'm advocating you do exactly the same for `__eq__` as you did for the others

Reply:
Version sets are infinite, so I'm not sure if getting this working in a Python hash set is a requirement. Lets decide what we want the end product to look like and go from there.

Reply:
Well, even without doing that, there are very few python objects whose `__eq__` throws like this - comparing `icarus_ver == modelsim_ver` is far less nonsensical than `"three" == 3`, but the latter still doesn't error.

Reply:
I'm not against it returning `False`, but probing for a direction from anybody. Python's behavior here is inconsistent simply to better support how sets are implemented. The other ordering operations throw. I guess consistently inconsistent is preferable to inconsistently consistent.

Reply:
Note that in python 2, `3 < "three"` was legal albeit giving a consistent but otherwise mostly meaningless result.